[김현진] 2장. 렌더링 코드 적용

guswls1846 opened the pull request:
## 변경 내용
- `프레임워크 없는 프론트엔드 개발 2장. 렌더링`의 내용을 적용하기 위해 기존 웹 컴포넌트로 만들어져 있던  `<todo-app>`을 주석 처리하고  `data-component`를 이용한 가상돔 동적 렌더링 방식으로 변경했습니다. 
- 책과 진도를 맞추기 위해 이벤트 리스너를 이용한 기능 구현을 제외 했습니다.
- 책에 있는 예제코드를 미약하지만 TypeScript를 적용했습니다.
- 책에 있는 `faker-js` 코드를 그대로 사용하면 deprecated되는게 많아서 관련 코드를 최신 버전에 맞게 문법을 조정했습니다.
- vite에서 jest 환경을 구성하고 코파일럿의 도움으로 약간의 테스트 코드를 만들어 봤습니다.  [jest 환경설정을 위한 참고 블로그](https://velog.io/@a770418/Vite-TypeScript-%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C-babel%EC%9D%84-%EC%9D%B4%EC%9A%A9%ED%95%9C-jest-%EC%A0%81)

Commit:
build: @faker-js/faker 설치

Commit:
feat: getTodos함수 생성

- faker js 라이브러리를 사용하여 랜덤 todo 리스트를 생성해주는 함수입니다.

Commit:
feat: State, Component, Registry 공통 타입 생성

Commit:
feat: registry 함수 생성

- 변경점이 생기면 자동으로 렌더링 해주는 함수 입니다.

Commit:
modify: RegistryKey 타입 생성

- 렌더링 되어야할 registry를 추적하기 위해 타입을 생성합니다.

Commit:
feat: 간단한 diff알고리즘 함수 생성

Commit:
modify: 웹 컴포넌트로 감싸져있던 todo-app 템플릿 코드 풀기

- 가상돔을 이용한 동적 데이터 렌더링을 구현하기 위해 웹컴포넌트를 임시적으로 주석 처리합니다.

Commit:
feat: todo-app css코드 한 곳 에 모은 파일 생성

Commit:
feat: todos widget 생성

Commit:
feat: filters 위젯 생성

Commit:
feat: counter 위젯 생성

Commit:
feat: provider 함수 생성

- 가상돔 렌더링을 실행하는 함수입니다.

jjanmo commented on hyunjin/client/package.json:
```diff
@@ -9,6 +9,7 @@
     "preview": "vite preview"
   },
   "devDependencies": {
+    "@faker-js/faker": "^8.4.1",
```
👍🏻 많이 바꼈더라구요 ㅎㅎ
저는 책 코드 다운받아서 변형해봤는데... 옛날버전CDN으로는 돌아지 않더군요 후후

jjanmo commented on hyunjin/client/src/shared/utils/getTodos.ts:
```diff
@@ -0,0 +1,21 @@
+import { Todo } from "@shared/types";
+
+const createElement = () => ({
+  text: faker.lorem.words(),
```
(궁금) words() 를 빈칸이면 기본값이 몇인가요??

guswls1846 replied:
>(궁금) words() 를 빈칸이면 기본값이 몇인가요??

@jjanmo 뜻을 알수 없는 랜덤 단어가 생성됩니다. 라이브러리에 내제되어 있는 [메타데이터](https://github.com/faker-js/faker/blob/8134b336f3c270126b5f8af9f18a7e7563c104bd/src/locales/en/word/adjective.ts)가 랜덤으로 반환되는거 같아요

jjanmo commented on hyunjin/client/src/widgets/filters/ui/index.ts:
```diff
@@ -0,0 +1,15 @@
+
+export const Filters = (targetElement: Element, { currentFilter }: State) => {
+  const newCounter = targetElement.cloneNode(true) as Element;
+  Array.from(newCounter.querySelectorAll(".filter-list")).forEach((a) => {
```
이건 개인적인 부분이지만, 실제 책 코드에서 `a` 태그를 말하는거라서 `a` 라고 쓴거겠지만, `aElem` 이나 `anchor` 등등의 구체적인 네이밍이 좋다고 생각하거든요. 속으로 구체적으로 써주지 하는 생각을 했습니다 😅

guswls1846 replied:
9ed4cde
그러네요 책에서는 부모요소를 `li a`로 잡고 있어서 `.filter-list` 클래스명으로 바꾼건데 자식이름도 너무 성의 없이 지었네요

jjanmo reviewed (approved):
고생하셨습니다 👍🏻

Commit:
build: jest 환경 구성

- 참조 블로그:https://velog.io/@a770418/Vite-TypeScript-%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C-babel%EC%9D%84-%EC%9D%B4%EC%9A%A9%ED%95%9C-jest-%EC%A0%81

Commit:
test: counter 테스트 코드 추가

Commit:
test: filter 테스트 코드 추가

Commit:
test: todos 테스트 코드 추가

Commit:
refactor: 이름 수정

rtchamp reviewed (commented):
jest는 어떤가요?

실무하실때 테스트 프레임웍 많이들 쓰시나요?

ghost commented:
> jest는 어떤가요?
> 
> 실무하실때 테스트 프레임웍 많이들 쓰시나요?

저도 실무를 하면서 팀원이 협업하여 테스트 코드를 작성해본 경험은 없습니다.  우대 사항으로 테스트 코드 작성 경험이 있는 회사에 가서 면접을 봤을때도 실제 테스트 코드를 작성하는 회사는 없었고요.  제 짧은 경험을 공유드리자면 규모가 크고 문화가 잡힌 회사에서는 프론트엔드도tc가 필수로 적용되는 곳도 있지만 부서바이 부서, 팀바이 팀, 규모가 작은 회사, 이제 갓 시작한 스타트업, 프리랜서 등등은 아직까지도 테스트코드의 필요성을 인지하면서도 작성은 하지 않는 듯 합니다. 

그래도 최근에는 테스트 코드를 ai가 작성해줘서 진입장벽이 많이 낮아진거 같긴해요!
